Src/VimCore/CommonOperations.fs

```diff
@@ -2387,19 +2427,385 @@ type internal CommonOperations
+                let column = x.GetAppropriateColumnForSpaces line spaces
+                column.VirtualStartPoint
+
+            let caretPoint = getRelativePoint primarySelectedSpan.CaretPoint
+            let anchorPoint = getRelativePoint primarySelectedSpan.AnchorPoint
+            let activePoint = getRelativePoint primarySelectedSpan.ActivePoint
+            SelectedSpan(caretPoint, anchorPoint, activePoint)
+            |> x.AddSelectedSpan
+
+        // Choose an appropriate line to add the caret on.
+        match direction with
+        | Direction.Up ->
+            let firstLine =
+                selectedSpans.[0].CaretPoint
+                |> VirtualSnapshotPointUtil.GetContainingLine
+            if firstLine.LineNumber > 0 then
+                addSelectionOnLineNumber (firstLine.LineNumber - 1)
+        | Direction.Down ->
+            let lastLine =
+                selectedSpans.[selectedSpans.Length - 1].CaretPoint
+                |> VirtualSnapshotPointUtil.GetContainingLine
+            let lastLineNumber = SnapshotUtil.GetLastNormalizedLineNumber lastLine.Snapshot
+            if lastLine.LineNumber < lastLineNumber then
+                addSelectionOnLineNumber (lastLine.LineNumber + 1)
+        | _ ->
+            ()
+
+    /// Run the specified action for all selections
+    member x.RunForAllSelections action =
+
+        // Unless there are multiple selections just do the action once and
+        // return its result normally.
+        if x.IsMultiSelectionSupported then
+            let selectedSpans = x.SelectedSpans |> Seq.toList
+            if selectedSpans.Length > 1 then
+                x.RunForAllSelectionsCore action selectedSpans
+            else
+                action()
         else
-            TrackingPointUtil.GetPointInSnapshot point pointTrackingMode snapshot
-            |> OptionUtil.getOrDefault (SnapshotPoint(snapshot, min point.Position snapshot.Length))
+            action()
+
+    /// Run the specified action for the specified selected spans
+    member x.RunForAllSelectionsCore action selectedSpans =
+
+        // Get the current kind of visual mode, if any.
+        let visualModeKind =
+            _vimTextBuffer.ModeKind
+            |> VisualKind.OfModeKind
+
+        // Get any mode argument from the specified command result.
+        let getModeArgument result =
+            match result with
+            | CommandResult.Completed modeSwitch ->
+                match modeSwitch with
+                | ModeSwitch.SwitchModeWithArgument (_, modeArgument) ->
+                    Some modeArgument
+                | _ ->
+                    None
+            | _ ->
+                None
+
+        // Get any linked transaction from the specified command result.
+        let getLinkedTransaction result =
+            match getModeArgument result with
+            | Some (ModeArgument.InsertWithTransaction linkedTransaction) ->

```

Comment:
`InsertWithCountAndNewLine` and `InsertBlock` also has a linked transaction here. Consider just adding a member `LinkedUndoTransaction` to `ModeArgument`